lib/test-collection.js

```diff
@@ -165,26 +165,26 @@ class TestCollection extends EventEmitter {
 		return test;
 	}
 
-	_buildTestWithHooks(test) {
+	_buildTestWithHooks(test, context) {
 		if (test.metadata.skipped || test.metadata.todo) {
 			return new Sequence([this._skippedTest(this._buildTest(test))], true);
 		}
 
-		const context = {context: {}};
+		const contextRef = context ? Object.create(context) : {context: {}};
```

Comment:
It's actually `context.context` that needs to be (shallowly) cloned. Let's use <https://www.npmjs.com/package/lodash.clone>.

You'll have to create a context ref for the `before` and `after` hooks, and *before* those hooks have run you need to pass a cloned ref to the `beforeEach` and `afterEach` hooks as well as the tests. Something like this should work if you [modify the `t.context` getter and setter](https://github.com/avajs/ava/blob/a051d3e18dba92c893fddb08490a8627f586c231/lib/test.js#L55:L69).

```js
class ContextRef {
  constructor () {
    this.value = {}
  }

  get () {
    return this.value
  }

  set (newValue) {
    this.value = newValue
  }

  copy () {
    return new LateBinding(this)
  }
}

class LateBinding extends ContextRef {
  constructor (ref) {
    super()
    this.ref = ref
    this.bound = false
  }

  get () {
    if (!this.bound) this.set(clone(this.ref.get()))
    return super.get()
  }

  set (newValue) {
    this.bound = true
    super.set(newValue)
  }
}
```

The same copied ref should be passed to the `beforeEach` and `afterEach` hooks as well as the tests. The first time the context is accessed it'll be "bound" to the value set in the `before` hooks. You can pass the original ref to the `after` hooks.

Reply:
@novemberborn , this is quite circular, and gives me an error on XO, I now understand that the `contextRef` is injected into the `Test` class then to the `ExecutionContext` (that is why you pointed me that way), is it necessary that we should create the `LateBinding` class? Is it possible that we can just return the `copy()` method a clone of itself?